Slightly flaky test DatagramChannelTests.testEcnSendReceiveIPV4

Discovered via internal tools

```
Test Suite 'Selected tests' started at 2022-05-13 16:28:40.627
Test Suite 'DatagramChannelTests' started at 2022-05-13 16:28:40.629
Test Case 'DatagramChannelTests.testEcnSendReceiveIPV4' started at 2022-05-13 16:28:40.629
/workspace/.logs/FA64A58E-D8A3-40BE-A7B9-8C06931AD7D6/repos/swift-nio/Tests/NIOPosixTests/DatagramChannelTests.swift:723: error: DatagramChannelTests.testEcnSendReceiveIPV4 : XCTAssertEqual failed: ("Optional(NIOCore.NIOExplicitCongestionNotificationState.congestionExperienced)") is not equal to ("Optional(NIOCore.NIOExplicitCongestionNotificationState.transportNotCapable)") - 
/workspace/.logs/FA64A58E-D8A3-40BE-A7B9-8C06931AD7D6/repos/swift-nio/Tests/NIOPosixTests/DatagramChannelTests.swift:723: error: DatagramChannelTests.testEcnSendReceiveIPV4 : XCTAssertEqual failed: ("Optional(NIOCore.NIOExplicitCongestionNotificationState.transportCapableFlag0)") is not equal to ("Optional(NIOCore.NIOExplicitCongestionNotificationState.congestionExperienced)") - 
/workspace/.logs/FA64A58E-D8A3-40BE-A7B9-8C06931AD7D6/repos/swift-nio/Tests/NIOPosixTests/DatagramChannelTests.swift:723: error: DatagramChannelTests.testEcnSendReceiveIPV4 : XCTAssertEqual failed: ("Optional(NIOCore.NIOExplicitCongestionNotificationState.transportCapableFlag1)") is not equal to ("Optional(NIOCore.NIOExplicitCongestionNotificationState.transportCapableFlag0)") - 
/workspace/.logs/FA64A58E-D8A3-40BE-A7B9-8C06931AD7D6/repos/swift-nio/Tests/NIOPosixTests/DatagramChannelTests.swift:723: error: DatagramChannelTests.testEcnSendReceiveIPV4 : XCTAssertEqual failed: ("Optional(NIOCore.NIOExplicitCongestionNotificationState.transportNotCapable)") is not equal to ("Optional(NIOCore.NIOExplicitCongestionNotificationState.transportCapableFlag1)") - 
Test Case 'DatagramChannelTests.testEcnSendReceiveIPV4' failed (0.009 seconds)
Test Suite 'DatagramChannelTests' failed at 2022-05-13 16:28:40.638
	 Executed 1 test, with 4 failures (0 unexpected) in 0.009 (0.009) seconds
Test Suite 'Selected tests' failed at 2022-05-13 16:28:40.638
	 Executed 1 test, with 4 failures (0 unexpected) in 0.009 (0.009) seconds
Command failed reason=exit, status=1
```

--- a/Tests/NIOPosixTests/DatagramChannelTests.swift
+++ b/Tests/NIOPosixTests/DatagramChannelTests.swift
@@ -719,11 +719,16 @@ final class DatagramChannelTests: XCTestCase {
             let expectedNumReads = ecnStates.count * (vectorSend ? 2 : 1)
             let reads = try receiveChannel.waitForDatagrams(count: expectedNumReads)
             XCTAssertEqual(reads.count, expectedNumReads)
-            for read in reads {
-                // Datagrams may be received out-of-order, let's find the associated ecnStateIdx.
-                let ecnStateIdx: Int = try XCTUnwrap(read.data.getInteger(at: read.data.readerIndex))
-                XCTAssertEqual(read.metadata?.ecnState, ecnStates[ecnStateIdx])
-                XCTAssertEqual(read.metadata?.packetInfo, expectedPacketInfo)
+            for (ecnStateIdx, ecnState) in ecnStates.enumerated() {
+                // Datagrams may be received out-of-order, let's find the associated reads.
+                let encStateReads = reads.filter { read in
+                    read.data.getInteger(at: read.data.readerIndex) == ecnStateIdx
+                }
+                XCTAssertEqual(encStateReads.count, vectorSend ? 2 : 1)
+                for read in encStateReads {
+                    XCTAssertEqual(read.metadata?.ecnState, ecnState)
+                    XCTAssertEqual(read.metadata?.packetInfo, expectedPacketInfo)
+                }
             }
         } ())
     }

--- a/Tests/NIOPosixTests/DatagramChannelTests.swift
+++ b/Tests/NIOPosixTests/DatagramChannelTests.swift
@@ -706,11 +706,7 @@ final class DatagramChannelTests: XCTestCase {
                 // Datagrams may be received out-of-order, so we use the payload to associate the read.
                 let writeData = AddressedEnvelope(
                     remoteAddress: receiveChannel.localAddress!,
-                    data: sendChannel.allocator.buffer(
-                        integer: ecnStateIdx,
-                        endianness: .big,
-                        as: Int.self
-                    ),
+                    data: sendChannel.allocator.buffer(integer: ecnStateIdx),
                     metadata: .init(ecnState: ecnState, packetInfo: expectedPacketInfo)
                 )
                 // Sending extra data without flushing should trigger a vector send.
@@ -725,11 +721,7 @@ final class DatagramChannelTests: XCTestCase {
             XCTAssertEqual(reads.count, expectedNumReads)
             for read in reads {
                 // Datagrams may be received out-of-order, let's find the associated ecnStateIdx.
-                let ecnStateIdx = try XCTUnwrap(read.data.getInteger(
-                    at: 0,
-                    endianness: .big,
-                    as: Int.self
-                ))
+                let ecnStateIdx: Int = try XCTUnwrap(read.data.getInteger(at: read.data.readerIndex))
                 XCTAssertEqual(read.metadata?.ecnState, ecnStates[ecnStateIdx])
                 XCTAssertEqual(read.metadata?.packetInfo, expectedPacketInfo)
             }

--- a/Tests/NIOPosixTests/DatagramChannelTests.swift
+++ b/Tests/NIOPosixTests/DatagramChannelTests.swift
@@ -702,34 +702,32 @@ final class DatagramChannelTests: XCTestCase {
                                                                        .congestionExperienced,
                                                                        .transportCapableFlag0,
                                                                        .transportCapableFlag1]
-            for (ecnStateIdx, ecnState) in ecnStates.enumerated() {
-                // Datagrams may be received out-of-order, so we use the payload to associate the read.
+            // Datagrams may be received out-of-order, so we use a sequential integer in the payload.
+            let metadataWrites: [(Int, AddressedEnvelope<ByteBuffer>.Metadata?)] = try ecnStates.enumerated().reduce(into: []) { metadataWrites, ecnState in
                 let writeData = AddressedEnvelope(
                     remoteAddress: receiveChannel.localAddress!,
-                    data: sendChannel.allocator.buffer(integer: ecnStateIdx),
-                    metadata: .init(ecnState: ecnState, packetInfo: expectedPacketInfo)
+                    data: sendChannel.allocator.buffer(integer: ecnState.offset),
+                    metadata: .init(ecnState: ecnState.element, packetInfo: expectedPacketInfo)
                 )
                 // Sending extra data without flushing should trigger a vector send.
                 if (vectorSend) {
                     sendChannel.write(writeData, promise: nil)
+                    metadataWrites.append((ecnState.offset, writeData.metadata))
                 }
                 try sendChannel.writeAndFlush(writeData).wait()
+                metadataWrites.append((ecnState.offset, writeData.metadata))
             }
 
             let expectedNumReads = ecnStates.count * (vectorSend ? 2 : 1)
-            let reads = try receiveChannel.waitForDatagrams(count: expectedNumReads)
-            XCTAssertEqual(reads.count, expectedNumReads)
-            for (ecnStateIdx, ecnState) in ecnStates.enumerated() {
-                // Datagrams may be received out-of-order, let's find the associated reads.
-                let encStateReads = reads.filter { read in
-                    read.data.getInteger(at: read.data.readerIndex) == ecnStateIdx
-                }
-                XCTAssertEqual(encStateReads.count, vectorSend ? 2 : 1)
-                for read in encStateReads {
-                    XCTAssertEqual(read.metadata?.ecnState, ecnState)
-                    XCTAssertEqual(read.metadata?.packetInfo, expectedPacketInfo)
-                }
+            let metadataReads = try receiveChannel.waitForDatagrams(count: expectedNumReads).map {
+                ($0.data.getInteger(at: $0.data.readerIndex, as: Int.self)!, $0.metadata)
             }
+
+            // Datagrams may be received out-of-order, so we order reads and writes by payload.
+            XCTAssertEqual(
+                metadataReads.sorted { $0.0 < $1.0 }.map { $0.1 },
+                metadataWrites.sorted { $0.0 < $1.0 }.map { $0.1 }
+            )
         } ())
     }
 

--- a/Tests/NIOPosixTests/DatagramChannelTests.swift
+++ b/Tests/NIOPosixTests/DatagramChannelTests.swift
@@ -698,16 +698,21 @@ final class DatagramChannelTests: XCTestCase {
                 XCTAssertNoThrow(try sendChannel.close().wait())
             }
             
-            var buffer = sendChannel.allocator.buffer(capacity: 1)
-            buffer.writeRepeatingByte(0, count: 1)
             let ecnStates: [NIOExplicitCongestionNotificationState] = [.transportNotCapable,
                                                                        .congestionExperienced,
                                                                        .transportCapableFlag0,
                                                                        .transportCapableFlag1]
-            for ecnState in ecnStates {
-                let writeData = AddressedEnvelope(remoteAddress: receiveChannel.localAddress!,
-                                                  data: buffer,
-                                                  metadata: .init(ecnState: ecnState, packetInfo: expectedPacketInfo))
+            for (ecnStateIdx, ecnState) in ecnStates.enumerated() {
+                // Datagrams may be received out-of-order, so we use the payload to associate the read.
+                let writeData = AddressedEnvelope(
+                    remoteAddress: receiveChannel.localAddress!,
+                    data: sendChannel.allocator.buffer(
+                        integer: ecnStateIdx,
+                        endianness: .big,
+                        as: Int.self
+                    ),
+                    metadata: .init(ecnState: ecnState, packetInfo: expectedPacketInfo)
+                )
                 // Sending extra data without flushing should trigger a vector send.
                 if (vectorSend) {
                     sendChannel.write(writeData, promise: nil)
@@ -715,12 +720,17 @@ final class DatagramChannelTests: XCTestCase {
                 try sendChannel.writeAndFlush(writeData).wait()
             }
 
-            let expectedReads = ecnStates.count * (vectorSend ? 2 : 1)
-            let reads = try receiveChannel.waitForDatagrams(count: expectedReads)
-            XCTAssertEqual(reads.count, expectedReads)
-            for readNumber in 0..<reads.count {
-                let read = reads[readNumber]
-                XCTAssertEqual(read.metadata?.ecnState, ecnStates[readNumber / (vectorSend ? 2 : 1)])
+            let expectedNumReads = ecnStates.count * (vectorSend ? 2 : 1)
+            let reads = try receiveChannel.waitForDatagrams(count: expectedNumReads)
+            XCTAssertEqual(reads.count, expectedNumReads)
+            for read in reads {
+                // Datagrams may be received out-of-order, let's find the associated ecnStateIdx.
+                let ecnStateIdx = try XCTUnwrap(read.data.getInteger(
+                    at: 0,
+                    endianness: .big,
+                    as: Int.self
+                ))
+                XCTAssertEqual(read.metadata?.ecnState, ecnStates[ecnStateIdx])
                 XCTAssertEqual(read.metadata?.packetInfo, expectedPacketInfo)
             }
         } ())

--- a/Tests/NIOPosixTests/DatagramChannelTests.swift
+++ b/Tests/NIOPosixTests/DatagramChannelTests.swift
@@ -718,7 +718,7 @@ final class DatagramChannelTests: XCTestCase {
                 metadataWrites.append((ecnState.offset, writeData.metadata))
             }
 
-            let expectedNumReads = ecnStates.count * (vectorSend ? 2 : 1)
+            let expectedNumReads = metadataWrites.count
             let metadataReads = try receiveChannel.waitForDatagrams(count: expectedNumReads).map {
                 ($0.data.getInteger(at: $0.data.readerIndex, as: Int.self)!, $0.metadata)
             }

